app/routers/weather_forecast.py

```diff
@@ -0,0 +1,232 @@
+
+SUCCESS_STATUS = 0
+ERROR_STATUS = -1
+MIN_HISTORICAL_YEAR = 1975
+MAX_FUTURE_YEAR = 2050
+HISTORY_TYPE = "history"
+HISTORICAL_FORECAST_TYPE = "historical-forecast"
+FORECAST_TYPE = "forecast"
+INVALID_DATE_INPUT = "Invalid date input provided"
+INVALID_YEAR = "Year is out of supported range"
+HISTORY_URL = "https://visual-crossing-weather.p.rapidapi.com/history"
+FORECAST_URL = "https://visual-crossing-weather.p.rapidapi.com/forecast"
+HEADERS = {'x-rapidapi-host': "visual-crossing-weather.p.rapidapi.com"}
+BASE_QUERY_STRING = {"aggregateHours": "24", "unitGroup": "metric", "dayStartTime": "00:00:01",
+                     "contentType": "json", "dayEndTime": "23:59:59", "shortColumnNames": "True"}
+HISTORICAL_AVERAGE_NUM_OF_YEARS = 3
+NO_API_RESPONSE = "No response from server"
+
+
+def validate_date_input(requested_date):
+    """ date validation.
+    Args:
+        requested_date (date) - date requested for forecast.
+    Returns:
+        (bool) - validate ended in success or not.
+        (str) - error message.
+    """
+    if isinstance(requested_date, datetime.date):
+        if MIN_HISTORICAL_YEAR <= requested_date.year <= MAX_FUTURE_YEAR:
+            return True, None
+        else:
+            return False, INVALID_YEAR
+    else:
+        return False, INVALID_DATE_INPUT
+
+
+def get_data_from_weather_api(url, input_query_string):
+    """ get the relevant weather data by calling the "Visual Crossing Weather" API.
+    Args:
+        url (str) - API url.
+        input_query_string (dict) - input for the API.
+    Returns:
+        (json) - JSON data returned by the API.
+        (str) - error message.
+    """
+    load_dotenv()
+    HEADERS['x-rapidapi-key'] = os.getenv('WEATHER_API_KEY')
+    try:
+        response = requests.request("GET", url, headers=HEADERS, params=input_query_string)
+    except requests.exceptions.RequestException:
+        return None, NO_API_RESPONSE
+    if response:
+        try:
+            return response.json()["locations"], None
+        except KeyError:
+            return None, response.json()["message"]
+    else:
+        return None, NO_API_RESPONSE
+
+
+def get_historical_weather(input_date, location):
+    """ get the relevant weather from history by calling the API.
+    Args:
+        input_date (date) - date requested for forecast.
+        location (str) - location name.
+    Returns:
+        weather_data (json) - output weather data.
+        error_text (str) - error message.
+    """
+    input_query_string = BASE_QUERY_STRING
+    input_query_string["startDateTime"] = input_date.isoformat()
+    input_query_string["endDateTime"] = (input_date + datetime.timedelta(days=1)).isoformat()
+    input_query_string["location"] = location
+    api_json, error_text = get_data_from_weather_api(HISTORY_URL, input_query_string)
+    location_found = list(api_json.keys())[0]
+    if api_json:
+        weather_data = {'MinTempCel': api_json[location_found]['values'][0]['mint'],
```

Comment:
Can we be absolutely sure there is no `['values']` list with 0 items in it?

Reply:
In this phase we are after finding the "location" tag in the json, if the API had any problems we would have found them earlier. I don't think a try statement is required here.